include/perfetto/ext/base/flat_hash_map.h

```diff
@@ -264,13 +265,19 @@ class FlatHashMap {
     if (PERFETTO_UNLIKELY(capacity_ == 0))
       return;
 
-    for (size_t i = 0; i < capacity_; ++i) {
-      const uint8_t tag = tags_[i];
-      if (tag != kFreeSlot && tag != kTombstone)
-        EraseInternal(i);
+    if (!std::is_trivially_destructible<Key>::value ||
+        !std::is_trivially_destructible<Value>::value) {
+      for (size_t i = 0; i < capacity_; ++i) {
+        const uint8_t tag = tags_[i];
+        if (tag != kFreeSlot && (AppendOnly || tag != kTombstone)) {
+          keys_[i].~Key();
+          values_[i].~Value();
+        }
+      }
     }
-    // Clear all tombstones. We really need to do this for AppendOnly.
-    MaybeGrowAndRehash(/*grow=*/false);
+    memset(&tags_[0], kFreeSlot, capacity_);
```

Comment:
could we call Reset(capacity_) here? does it make a big difference?
I feel it keeps the code more mainteinable.

maybe you can add, in the Reset() a branch like:

size_t old_capacity = capacity_;
... (current reset code)
if (n != old_capacity) {
    keys_ = AlignedAllocTyped<Key[]>(n);      // Deliberately not 0-initialized.
    values_ = AlignedAllocTyped<Value[]>(n);  // Deliberately not 0-initialized.
}

so it doesn't re-allocate.